File: parser/src/main/java/dev/mvmo/sicklang/parser/ast/expression/CallExpressionNode.java
package dev.mvmo.sicklang.parser.ast.expression;

import dev.mvmo.sicklang.token.Token;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;

import java.util.List;
import java.util.stream.Collectors;

@Getter
@Setter
@RequiredArgsConstructor(staticName = "newInstance")
public class CallExpressionNode implements ExpressionNode {

    private final Token token;
    private final ExpressionNode function; // identifier or function
    private List<ExpressionNode> arguments;

    @Override
    public String tokenLiteral() {
        return token.literal();
    }

    @Override
    public String toString() {
        return function.toString() +
                "(" +
                arguments.stream().map(ExpressionNode::toString).collect(Collectors.joining(", ")) +
                ")";
    }

}

File: parser/src/main/java/dev/mvmo/sicklang/parser/ast/expression/CallExpressionNode.kt
package dev.mvmo.sicklang.parser.ast.expression

import dev.mvmo.sicklang.token.Token

class CallExpressionNode(val token: Token, val function: ExpressionNode) : ExpressionNode {

    var arguments: List<ExpressionNode>? = null

    override fun tokenLiteral(): String =
        token.literal

    override fun toString(): String =
        "$function(${arguments?.joinToString(transform = ExpressionNode::toString) ?: ""})"

}